crates/harp/src/vector/list.rs

```diff
@@ -0,0 +1,160 @@
+//
+// list.rs
+//
+// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+//
+//
+
+use libr::SEXP;
+
+use crate::object::list_cbegin;
+use crate::object::r_length;
+use crate::object::r_list_poke;
+use crate::object::RObject;
+use crate::r_typeof;
+
+pub struct List {
+    inner: RObject,
+    ptr: *const SEXP,
+}
+
+pub struct ListIter {
+    index: usize,
+    size: usize,
+    ptr: *const SEXP,
+    _shelter: RObject,
+}
+
+impl List {
+    pub fn iter(&self) -> ListIter {
+        unsafe { ListIter::new_unchecked(self.inner.sexp) }
+    }
+}
+
+impl super::Vector for List {
+    type Type = RObject;
+    type Item = SEXP;
+    const SEXPTYPE: u32 = libr::VECSXP;
+    type UnderlyingType = SEXP;
+    type CompareType = RObject;
+
+    unsafe fn new_unchecked(object: impl Into<SEXP>) -> Self {
+        let object: SEXP = object.into();
+        let ptr = crate::list_cbegin(object);
+
+        Self {
+            inner: object.into(),
+            ptr,
+        }
+    }
+
+    fn data(&self) -> SEXP {
+        self.inner.sexp
+    }
+
+    // Never missing. We can't treat `NULL` as missing because it would cause
+    // getters to return `None` or a `MissingValueError`.
+    fn is_na(_x: &Self::UnderlyingType) -> bool {
+        false
+    }
+
+    fn get_unchecked_elt(&self, index: isize) -> Self::UnderlyingType {
+        unsafe { *self.ptr.wrapping_add(index as usize) }
+    }
+
+    fn convert_value(x: &Self::UnderlyingType) -> Self::Type {
+        (*x).into()
+    }
+
+    unsafe fn create<T>(data: T) -> Self
+    where
+        T: IntoIterator,
+        <T as IntoIterator>::IntoIter: ExactSizeIterator,
+        <T as IntoIterator>::Item: AsRef<Self::Item>,
+    {
+        let mut data = data.into_iter();
+
+        let size = data.len();
+        let inner = crate::alloc_list(size).unwrap();
+        let inner: RObject = inner.into();
+
+        for i in 0..size {
+            unsafe {
+                let value = data.next().unwrap_unchecked();
+                let value = value.as_ref();
+                r_list_poke(inner.sexp, i as libr::R_xlen_t, *value)
```

Comment:
```suggestion
                r_list_poke(inner.sexp, i as libr::R_xlen_t, *value)
```

Use conversion helper for `i as libr::R_xlen_t`? Not sure how strict to be

Reply:
Going from unsized to sized should not fail so this seems fine? Not sure either.